feat: Sends an Overlay Message Whenever You Gain Combat XP

san7890 opened the pull request:
![image](https://user-images.githubusercontent.com/34697715/206893729-4affd527-e556-402d-aba0-a74fc35de280.png)
![image](https://user-images.githubusercontent.com/34697715/206893732-3f76e673-2ff6-4aad-bf35-62dfe6a46369.png)

This just ports over this Legacy feature (`GameUpdateEXPMessages`), with a lot of the code reworked to fit the new Artemis framework (as well as just be a lot easier to read/work with). It ran on every tick in Legacy, so I believe it should be fine to work here as well? Let me know if there's a better solution. A bunch of the math might be un-necessary, but it took me a while to get to something that I felt was appropriate.

Thanks for your time!

san7890 committed:
who knows if this will work

san7890 committed:
wait this should actually work

san7890 committed:
whoops

san7890 committed:
oh thank god it works

i was getting worried

san7890 committed:
forgot l18n and decimalformat fix

san7890 committed:
patches string format

san7890 committed:
maaath

san7890 committed:
final math cleanup

san7890 committed:
cull debug information

san7890 committed:
makes the message match legacy

san7890 committed:
spotless

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+import com.wynntils.wynn.utils.WynnUtils;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+@FeatureInfo
```
You only need this if you want to change defaults. So you either forgot to do that or just have this here unneeded

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+public class CombatXPGainMessageFeature extends UserFeature {
+
+    @Config
+    public boolean getCombatXPGainMessages = true;
```
In Artemis, we don't need configs like this. If you disable the feature, then you wont get combat xp gain messages :)

san7890 replied:
ah i misunderstood that the last time you told me about it

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+
+        if (newTickXP == lastTickXP) {
+            return;
+        }
```
```suggestion
        if (newTickXP == lastTickXP)  return;
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+    public void onTick(ClientTickEvent.End event) {
+        if (!WynnUtils.onWorld() || !getCombatXPGainMessages) {
+            return;
+        }
```
```suggestion
        if (!WynnUtils.onWorld()) return;
```

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static float newTickXP = 0;
+    private static float lastTickXP = 0;
+    private static int trackedPercentage = 0;
```
Why is this an int?

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+        }
+
+        int neededXP = data.getXpPointsNeededToLevelUp();
+        if (lastTickXP != 0) {
```
I am not sure if defaulting to 0 is beneificial, does this work nicely with level ups?

san7890 replied:
The only time when experience is 0 is at the very start of a session, when you join a world. Otherwise, it will be a positive integer by some degree. There might be an edge case regarding that where you could get exactly 0 points, but that'll smooth over by the next second. Not sure what the best way to handle that edge case is, or even it's important. 

I'll test it in-game with a levelup and see what comes of it to ensure nothing's breaking.

kristofbolyai replied:
But theoritically you could stay on 0 xp. I think. So maybe, use a different magic number for the edge case? -1?

san7890 replied:
If you "stay on" 0 XP then you get filtered by this above check:

```java
if (newTickXP == lastTickXP) return;
```

or do I misunderstand? This code only runs whenever there is an actual change in the XP from the last time we ran. The purpose for the if/else check here is only for the case for when you start at 0 XP upon joining the World, and then Wynncraft sends you your character's saved amount of XP within a few ticks of being in-world. Otherwise, you would get this type of message:

```txt
+ 50294 XP (0.0%)
```

Because, lastTickXP was 0. 0/Any Number is 0. We haven't set lastTickXP to anything since this method hasn't fully ran to the end where we actually give it a number to hold onto. This doesn't make any sense, so the stop-guard here is to account for that and ensure that you really get something like:

```txt
+ 50294 XP (54.2%)
```

Might be worthwhile to write a comment on that. Sorry I didn't clarify on this earlier, I just woke up a bit ago.

kristofbolyai replied:
oh. okay, perhaps this is fine

san7890 commented:
current iteration of the code has it run every tick because of a false assertion i made reading legacy code, whoops. i'll mark this ready for review when i'm able to fix that.

kristofbolyai commented on common/src/main/resources/assets/wynntils/lang/en_us.json:
```diff
@@ -171,6 +171,8 @@
   "feature.wynntils.gammabright.name": "Gammabright Keybind",
   "feature.wynntils.gearViewer.name": "Gear Viewer",
+  "feature.wynntils.getCombatXPGainMessages.description": "Do you want to see updates to your XP as overlay messages?",
+  "feature.wynntils.getCombatXPGainMessages.name": "Combat XP Gain Messages",
```
So you still don't use the run configurations in your IDE. Or you just didn't test this. But, you didn't add an i18n for the feature name, which should make it fail immediately in development environment.

kristofbolyai replied:
In fact, the i18n for the variable, which is not needed as I commented above, should not even look like this. You missed the whole feature part.

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,59 @@
+
+        NotificationManager.queueMessage(message);
+
+        lastTickXP = newTickXP;
```
Maybe set this before making a notification, so notification queuing is the last step of the process?

kristofbolyai replied:
(just to make it nicer imo)

kristofbolyai reviewed (commented):
Great feature, I missed this! However, there are a few problems. When you fix them up, this will be great!

san7890 committed:
implements configurable combat message tick delay

forgot about porting this from legacy

san7890 committed:
ci: spotless formatting

Incompleteusern committed:
Merge branch 'main' into xp-differential

san7890 committed:
Fixes up config and l18n

san7890 committed:
moves last tick XP before queueMessage

san7890 committed:
formatting of early returns

san7890 committed:
int -> float

cruft from earlier implementation

san7890 committed:
spotlessian

kristofbolyai commented:
Just to note this in this PR. A new event hooked into `ClientPacketListener#handleSetExperience` would be better than on tick.

san7890 committed:
lol it breaks if you're level 106

good thing i had a character at that level otherwise i never woulda caught this

san7890 commented:
> Just to note this in this PR. A new event hooked into `ClientPacketListener#handleSetExperience` would be better than on tick.

It would be nicer but you would have to handle tick-counting separately somehow to ensure grouping (and that you don't get flooded with a message on every instance of XP changing).

kristofbolyai commented:
> > Just to note this in this PR. A new event hooked into `ClientPacketListener#handleSetExperience` would be better than on tick.
> 
> 
> 
> It would be nicer but you would have to handle tick-counting separately somehow to ensure grouping (and that you don't get flooded with a message on every instance of XP changing).

You make ticks seconds, use timestamps

san7890 committed:
comments the lastTickXP != 0

san7890 committed:
wait why were these static

san7890 committed:
accidentally doing integer division

san7890 committed:
silly typo

san7890 committed:
handles changing world state

san7890 committed:
something broke

san7890 committed:
fix one bug, another shows itself

san7890 committed:
okay maybe this should fix it

i think this worked at some point but my code outpaced it

san7890 committed:
handles levelling up gracefully

it was busted, crap. i'm going to make a new character to further test this PR bleaugh

san7890 committed:
Merge branch 'main' into xp-differential

san7890 commented:
I believe I've now tested all of the cases and should now have code to handle everything.

san7890 committed:
ci: spotless formatting

magicus commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,101 @@
+
+        tickCounter = 0;
+
+        CharacterManager.CharacterInfo data = WynnUtils.getCharacterInfo();
```
I think it could be a bit clearer if everything below this were extracted into a separate method, i.e. the things that actually are done when the ticks align.

magicus commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,101 @@
+            percentChange = percentGained * 100;
+        } else {
+            percentChange = trackedPercentage * 100;
+        }
```
My gut feeling here is that a lot of this seems like stuff that CharacterManager should be responsible for. But I can't clearly say how that should be done, so maybe it can stay here. It all seems a bit convoluted, but your comments indicate that this is since the problem is more complex than it seems, so I guess it is okay.

san7890 replied:
It would be nice if it could be more abstract, yea- the only problem is that there are so many cases I ran into while working out the kinks of this system that there really does need to be some sort of context-dependent way to keep catching and handling odd quirks.

san7890 committed:
branches into new method

magicus committed:
Merge branch 'main' into xp-differential

magicus committed:
Merge branch 'main' into xp-differential

kristofbolyai committed:
Merge branch 'main' into xp-differential

kristofbolyai commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,105 @@
+
+        // You get division by zero errors when you're at the level cap (i.e. 106 in Wynncraft 2.0.1).
+        // This needs to be updated if the level cap is ever raised.
+        if (newLevel > 105) return;
```
I do not like this at all. It is one thing that it does not work on new versions, but it would be nice if we did not straight up crash the thing... The only reason why I accept this is that we do not crash since the feature will capture the crash and disable itself.

kristofbolyai committed:
fix: Use SetExperienceEvent

kristofbolyai committed:
fix: Handle max level and 0 xp gain better

kristofbolyai commented:
Hey @magicus and @san7890. I pushed 2 commits to address the issues I found to be annoying/blocking for release. I didn't want to take away the learning opportunity to code this such text book event/mixin, but I am sure san7890 will find some other event to add after they finish their finals season. Sorry about that.

I think, this is now in a state where we can merge this safely, after reviews.

kristofbolyai committed:
Merge branch 'main' into xp-differential

magicus commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,107 @@
+import com.wynntils.wynn.utils.WynnUtils;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+
+public class CombatXPGainMessageFeature extends UserFeature {
```
I think it should be `CombatXpGainMessageFeature`, and similarly for the variables below, `newTickXp`etc. Combining acronyms traditionally written in all-caps with CamelCase is always difficult, but the normal Java convention is to treat them as words and camel-case them ("Url"), and I think that's the best. That is also what we have done before wrt to XP.

magicus commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,107 @@
+    @Config
+    public float secondDelay = 0.5f;
+
+    private long lastExperienceDisplayTime = 0;
```
```suggestion
    private long lastXpDisplayTime = 0;
```

perhaps?

magicus commented on common/src/main/java/com/wynntils/features/user/CombatXPGainMessageFeature.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    @SubscribeEvent
+    public void onExperienceChange(SetExperienceEvent event) {
```
```suggestion
    public void onXpChange(SetXpEvent event) {
```

I think we should stick to XP for experience points. We could use "Experience" consistently as well, but that is not as clear and concise, I think. Mixing the two randomly is the least good solution.

magicus commented on common/src/main/java/com/wynntils/mc/mixin/ClientPacketListenerMixin.java:
```diff
@@ -157,6 +158,12 @@ private void handleSetPlayerTeamPacketPre(ClientboundSetPlayerTeamPacket packet,
         }
     }
 
+    @Inject(method = "handleSetExperience", at = @At("HEAD"))
```
```suggestion
    @Inject(method = "handleSetExperience", at = @At("RETURN"))
```

kristofbolyai committed:
fix: Fix up for reviews

kristofbolyai committed:
Merge branch 'main' into xp-differential

magicus commented:
Thank you @san7890 for the initial implementation, and thanks @kristofbolyai for the event-based version (and for doing the experience -> xp rename).
